os/kernel/binary_manager/compression/lzma/Alloc.c

```diff
@@ -0,0 +1,505 @@
+#endif
+
+void SetLargePageSize()
+{
+#ifdef _7ZIP_LARGE_PAGES
+	SIZE_T size;
+	GetLargePageMinimumP largePageMinimum = (GetLargePageMinimumP)
+											GetProcAddress(GetModuleHandle(TEXT("kernel32.dll")), "GetLargePageMinimum");
+	if (!largePageMinimum) {
+		return;
+	}
+	size = largePageMinimum();
+	if (size == 0 || (size & (size - 1)) != 0) {
+		return;
+	}
+	g_LargePageSize = size;
+#endif
+}
+
+void *BigAlloc(size_t size)
+{
+	if (size == 0) {
+		return NULL;
+	}
+
+	PRINT_ALLOC("Alloc-Big", g_allocCountBig, size, NULL);
+
+#ifdef _7ZIP_LARGE_PAGES
+	{
+		SIZE_T ps = g_LargePageSize;
+		if (ps != 0 && ps <= (1 << 30) && size > (ps / 2)) {
+			size_t size2;
+			ps--;
+			size2 = (size + ps) & ~ps;
+			if (size2 >= size) {
+				void *res = VirtualAlloc(NULL, size2, MEM_COMMIT | MEM_LARGE_PAGES, PAGE_READWRITE);
+				if (res) {
+					return res;
+				}
+			}
+		}
+	}
+#endif
+
+	return VirtualAlloc(NULL, size, MEM_COMMIT, PAGE_READWRITE);
+}
+
+void BigFree(void *address)
+{
+	PRINT_FREE("Free-Big", g_allocCountBig, address);
+
+	if (!address) {
+		return;
+	}
+	VirtualFree(address, 0, MEM_RELEASE);
+}
+
+#endif
+
+static void *SzAlloc(ISzAllocPtr p, size_t size)
+{
+	UNUSED_VAR(p);
+	return MyAlloc(size);
+}
+
+static void SzFree(ISzAllocPtr p, void *address)
+{
+	UNUSED_VAR(p);
+	MyFree(address);
+}
+const ISzAlloc g_Alloc = { SzAlloc, SzFree };
+
+static void *SzMidAlloc(ISzAllocPtr p, size_t size)
+{
+	UNUSED_VAR(p);
+	return MidAlloc(size);
+}
+
+static void SzMidFree(ISzAllocPtr p, void *address)
+{
+	UNUSED_VAR(p);
+	MidFree(address);
+}
+const ISzAlloc g_MidAlloc = { SzMidAlloc, SzMidFree };
+
+static void *SzBigAlloc(ISzAllocPtr p, size_t size)
+{
+	UNUSED_VAR(p);
+	return BigAlloc(size);
+}
+
+static void SzBigFree(ISzAllocPtr p, void *address)
+{
+	UNUSED_VAR(p);
+	BigFree(address);
+}
+const ISzAlloc g_BigAlloc = { SzBigAlloc, SzBigFree };
+
+/*
+  uintptr_t : <stdint.h> C99 (optional)
+            : unsupported in VS6
+*/
+
+#ifdef _WIN32
+typedef UINT_PTR UIntPtr;
+#else
+/*
+   typedef uintptr_t UIntPtr;
+ */
+typedef ptrdiff_t UIntPtr;
+#endif
+
+#define ADJUST_ALLOC_SIZE 0
+/*
+#define ADJUST_ALLOC_SIZE (sizeof(void *)-1)
+*/
+/*
+  Use (ADJUST_ALLOC_SIZE = (sizeof(void *)-1)), if
+     MyAlloc() can return address that is NOT multiple of sizeof(void *).
+*/
+
+/*
+#define MY_ALIGN_PTR_DOWN(p, align) ((void *)((char *)(p) - ((size_t)(UIntPtr)(p) & ((align) - 1))))
+*/
+#define MY_ALIGN_PTR_DOWN(p, align) ((void *)((((UIntPtr)(p)) & ~((UIntPtr)(align) - 1))))
+
+#define MY_ALIGN_PTR_UP_PLUS(p, align) MY_ALIGN_PTR_DOWN(((char *)(p) + (align) + ADJUST_ALLOC_SIZE), align)
+
+#if defined(_POSIX_C_SOURCE)
+#if (_POSIX_C_SOURCE >= 200112L) && !defined(_WIN32)
+#define USE_posix_memalign
+#endif
+#endif
+
+/*
+  This posix_memalign() is for test purposes only.
+  We also need special Free() function instead of free(),
+  if this posix_memalign() is used.
+*/
+
+/*
+static int posix_memalign(void **ptr, size_t align, size_t size)
```

Comment:
please remove it. if not needed.

Reply:
This is initial port.
Will modify files under lzma borrowed from 'LZMA SDK' in a later PR.

Reply:
ok